sql/core/src/test/resources/sql-tests/results/cast.sql.out

```diff
@@ -866,10 +866,10 @@ struct<>
 -- !query output
 org.apache.spark.SparkArithmeticException
 {
-  "errorClass" : "CANNOT_CHANGE_DECIMAL_PRECISION",
+  "errorClass" : "NUMERIC_VALUE_OUT_OF_RANGE",
   "sqlState" : "22005",
   "messageParameters" : {
-    "value" : "Decimal(compact, 10, 18, 6)",
+    "value" : "0.000010",
```

Comment:
Slightly confusing, how 0.000010 is related to `10.123`?

Reply:
@srielau Please, take a look at 9 lines above to the query:
```sql
select cast(interval '10.123' second as decimal(1, 0))
```

Reply:
That's something we need to fix.  There are two methods in Decimal: `changePrecision` will change the internal value in place, while `toPrecision` copies a new instance and check whether the new instance can be changed.
It is not related to this PR though.

Reply:
In the other way, we can identify the potential issues easier with this error message improvement.

Reply:
Can we create a JIRA ticket to track this? Hopefully cast decimal to string does not use the buggy `toPlainString` function.

Reply:
Why toPlainString is buggy?

Reply:
it should return `10.123`, right?

Reply:
@cloud-fan I explained this one in the comment above. It is because we use `changePrecision` on casting interval as decimal.

Reply:
and we should use `toPrecision` instead?

Reply:
I would prefer showing the original interval value in the message.

Reply:
I think if we know that `changePrecision()` has side effects, its wrapper w/ throw the exception is incorrect:
```scala
  private[this] def changePrecision(
      value: Decimal,
      decimalType: DecimalType,
      nullOnOverflow: Boolean): Decimal = {
    if (value.changePrecision(decimalType.precision, decimalType.scale)) {
      value
    } else {
      if (nullOnOverflow) {
        null
      } else {
        throw QueryExecutionErrors.cannotChangeDecimalPrecisionError(
          value, decimalType.precision, decimalType.scale, getContextOrNull())
      }
    }
  }
```
A possible fix is to create a copy of `value` before its modification by `value.changePrecision(...)` otherwise we always pass incorrect value to `throw QueryExecutionErrors.cannotChangeDecimalPrecisionError(...)`

Reply:
if that's easy to do efficiently, go ahead

Reply:
> A possible fix is to create a copy of value before its modification by value.changePrecision(...)

I would suggest using the original value string representation in the error. It could be string/integer/interval..

Reply:
>> but I am thinking of to not modify internal value longVal

> if that's easy to do efficiently, go ahead

Here is the PR https://github.com/apache/spark/pull/37649